File: backend/app/schemas/pricing.py
from pydantic import BaseModel, Field
from typing import List, Optional

class FactorInput(BaseModel):
    weather: Optional[str] = "none" # "none" | "rain" | "heatwave" | "storm"
    is_festival: Optional[bool] = False
    scheduled_hour: Optional[int] = None # 0 to 23 (e.g. 23 for night)
    day_of_week: Optional[str] = None # "monday" ... "sunday"
    demand_level: Optional[str] = "normal" # "low" | "normal" | "high" | "peak"
    is_urgent: Optional[bool] = False # Immediate 15-min dispatch

class PricingCalculateRequest(BaseModel):
    service_id: str # e.g. "cat-plumber" or "plumber"
    factors: Optional[FactorInput] = Field(default_factory=FactorInput)
    custom_base_rate: Optional[float] = None

class AppliedFactorItem(BaseModel):
    id: str
    factor_type: str # weather | festival | time | day | demand | urgency
    name: str
    description: str
    multiplier_weight: float
    surcharge_amount: float

class PricingCalculateResponse(BaseModel):
    service_name: str
    service_id: str
    base_price: float
    applied_factors: List[AppliedFactorItem]
    total_multiplier_raw: float
    total_multiplier_applied: float
    multiplier_cap: float
    multiplier_cap_enforced: bool
    surcharge: float
    worker_earning: float
    worker_payout_percent: float
    cooperative_fee: float
    cooperative_fee_percent: float
    welfare_contribution: float
    welfare_contribution_percent: float
    final_price: float
    base_floor_guaranteed: bool

class PriceRuleCreateRequest(BaseModel):
    factor_type: str # weather | festival | time | day | demand | urgency
    name: str
    description: str
    multiplier: float
    is_active: bool = True
    condition_value: Optional[str] = None

class PriceRuleUpdateRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    multiplier: Optional[float] = None
    is_active: Optional[bool] = None
    condition_value: Optional[str] = None

class PriceRuleResponse(BaseModel):
    id: str
    factor_type: str
    name: str
    description: str
    multiplier: float
    is_active: bool
    condition_value: Optional[str] = None
    created_at: str

class PricingConfigResponse(BaseModel):
    max_multiplier_cap: float
    worker_payout_percent: float
    cooperative_fee_percent: float
    welfare_contribution_percent: float
    minimum_wage_floor_enforced: bool

class PricingConfigUpdateRequest(BaseModel):
    max_multiplier_cap: Optional[float] = None
    worker_payout_percent: Optional[float] = None
    cooperative_fee_percent: Optional[float] = None
    welfare_contribution_percent: Optional[float] = None

File: src/services/pricing.ts
export interface FactorInput {
  weather?: 'none' | 'rain' | 'heatwave' | 'storm';
  is_festival?: boolean;
  scheduled_hour?: number;
  day_of_week?: string;
  demand_level?: 'low' | 'normal' | 'high' | 'peak';
  is_urgent?: boolean;
}

export interface PricingCalculateRequest {
  service_id: string;
  factors?: FactorInput;
  custom_base_rate?: number;
}

export interface AppliedFactorItem {
  id: string;
  factor_type: string;
  name: string;
  description: string;
  multiplier_weight: number;
  surcharge_amount: number;
}

export interface PricingCalculateResponse {
  service_name: string;
  service_id: string;
  base_price: number;
  applied_factors: AppliedFactorItem[];
  total_multiplier_raw: number;
  total_multiplier_applied: number;
  multiplier_cap: number;
  multiplier_cap_enforced: boolean;
  surcharge: number;

  worker_earning: number;
  worker_payout_percent: number;
  cooperative_fee: number;
  cooperative_fee_percent: number;
  welfare_contribution: number;
  welfare_contribution_percent: number;
  final_price: number;
  base_floor_guaranteed: boolean;
}


export interface PriceRule {
  id: string;
  factor_type: string;
  name: string;
  description: string;
  multiplier: number;
  is_active: boolean;
  condition_value?: string;
  created_at: string;
}

export interface PricingConfig {
  max_multiplier_cap: number;
  worker_payout_percent: number;
  cooperative_fee_percent: number;
  welfare_contribution_percent: number;
  minimum_wage_floor_enforced: boolean;
}

const API_BASE = import.meta.env.VITE_API_BASE_URL || 'https://ps-1-rtys.vercel.app';

export async function calculatePricing(request: PricingCalculateRequest): Promise<PricingCalculateResponse> {
  const res = await fetch(`${API_BASE}/pricing/calculate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request)
  });
  if (!res.ok) {
    throw new Error('Failed to calculate pricing');
  }
  return res.json();
}

export async function fetchPricingRules(): Promise<PriceRule[]> {
  const res = await fetch(`${API_BASE}/pricing/rules`);
  if (!res.ok) throw new Error('Failed to fetch pricing rules');
  return res.json();
}

export async function createPricingRule(rule: Omit<PriceRule, 'id' | 'created_at'>): Promise<PriceRule> {
  const res = await fetch(`${API_BASE}/pricing/rules`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(rule)
  });
  if (!res.ok) throw new Error('Failed to create pricing rule');
  return res.json();
}

export async function updatePricingRule(ruleId: string, rule: Partial<PriceRule>): Promise<PriceRule> {
  const res = await fetch(`${API_BASE}/pricing/rules/${ruleId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(rule)
  });
  if (!res.ok) throw new Error('Failed to update pricing rule');
  return res.json();
}

export async function deletePricingRule(ruleId: string): Promise<{ message: string }> {
  const res = await fetch(`${API_BASE}/pricing/rules/${ruleId}`, {
    method: 'DELETE'
  });
  if (!res.ok) throw new Error('Failed to delete pricing rule');
  return res.json();
}

export async function fetchPricingConfig(): Promise<PricingConfig> {
  const res = await fetch(`${API_BASE}/pricing/config`);
  if (!res.ok) throw new Error('Failed to fetch pricing configuration');
  return res.json();
}

export async function updatePricingConfig(config: Partial<PricingConfig>): Promise<PricingConfig> {
  const res = await fetch(`${API_BASE}/pricing/config`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(config)
  });
  if (!res.ok) throw new Error('Failed to update pricing configuration');
  return res.json();
}
